## 4절 카이제곱 검정
### 적합성 검정

In [3]:
# Q. A회사에서 출시되는 제품 공정결과의 예를 보고 색상과 관계없이 동일한 비율로 제품을 생산하는지 적합성 검정을 수행해보자
import numpy as np
color=['Black','Gold','Purple','Red','White']   # 색상(분석에 필요x)
counts=[423,304,274,205,294]    # 수량
expected=300    # 기대도수(동일한비율=1500/5)

# 적합성 검정 수행
from scipy.stats import chisquare
chisquare(f_obs=counts,f_exp=expected)

Power_divergenceResult(statistic=82.94, pvalue=4.14849046718008e-17)

### 동질성 검정

In [4]:
# Q. A회사의 성별에 따른 제품 선호도조사의 예를 보고 성별 간 선호도의 차이 검증을 위한 동질성 검정을 수행하자.
import pandas as pd
from scipy.stats import chi2_contingency
# 교차표에 맞게 데이터 프레임생성
obj={'Good':[400,350],'Bad':[350,800]}
cross=pd.DataFrame(obj)

In [7]:
# 동질성 검정 수행
chi,p,df,expected=chi2_contingency(cross)
print(chi)  # 카이제곱 통계량
print(p)    # p- 값
print(df)   # 자유도
print(expected) # 기대빈도

98.66917336693975
2.984113573360021e-23
1
[[296.05263158 453.94736842]
 [453.94736842 696.05263158]]


- p-value가 매우 작으므로  귀무가설 기각, 성별에따라 선호도의 비율이 다르다고 할수있다

### 독립성검정

In [9]:
# Q. A회사의 출시된 제품의 한달간 판매량의 예이다. 이를  통해 성별과 색상이 서로 관련있는지 알아보기 위해 독립성 검정을 수행해보자.
obj={'Black':[1620,2380],'Gold':[385,615],'Purple':[778,1230],'Red':[394,610],'White':[800,180]}
cross=pd.DataFrame(obj)

In [10]:
# 동질성 검정 수행
from scipy.stats import chi2_contingency
chi,p,df,expected = chi2_contingency(cross)
print(chi)  # 카이제곱 통계량
print(p)    # p- 값
print(df)   # 자유도
print(expected) # 기대빈도

626.2830361414021
3.1731601033660286e-134
4
[[1769.12811388  442.28202847  888.10231317  444.05115658  433.4363879 ]
 [2230.87188612  557.71797153 1119.89768683  559.94884342  546.5636121 ]]


- p-value가 매우 작으므로 귀무가설 기각, 두변수는 종속적이다. 즉, 성별과 색상은 서로 관련이 있다

## 5절 비모수 검정
### 스피어만 상관계수 검정

In [12]:
import pandas as pd
from sklearn.datasets import load_diabetes

diabetes=load_diabetes()
data=pd.DataFrame(diabetes.data,columns=diabetes.feature_names)
data.info() # 확인

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442 entries, 0 to 441
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     442 non-null    float64
 1   sex     442 non-null    float64
 2   bmi     442 non-null    float64
 3   bp      442 non-null    float64
 4   s1      442 non-null    float64
 5   s2      442 non-null    float64
 6   s3      442 non-null    float64
 7   s4      442 non-null    float64
 8   s5      442 non-null    float64
 9   s6      442 non-null    float64
dtypes: float64(10)
memory usage: 34.7 KB


In [13]:
# Q. 1절의 피어슨 상관계수 대신 스피어만 상관계수로 상관관계를 분석해보자.
# scipy.stats.speamanr
from scipy.stats import spearmanr
spearmanr(a=data['sex'],b=data['bmi'])

SignificanceResult(statistic=0.09807947297621517, pvalue=0.03929011358104615)

- 0.098로 상관계수가 매우 적은것으로 나타남
- p-value가 매우 작으므로 귀무가설을 기각 상관계수가 작지만 유의미하다고 할수있다

In [15]:
data[['sex','bmi']].corr(method='spearman')
# 으로도 확인가능

sex       bmi
sex  1.000000  0.098079
bmi  0.098079  1.000000

### 켄달의 타우 검정

In [16]:
# Q. 두 개의 등수 데이터를 임의로 생성하고 두 순위간 상관관계가 있는지 검정해보자.
import numpy as np
x=np.array([5,4,3,6,1,2])
y=np.array([1,5,2,2,2,6])

# 켄달의 타우 검정 실시
from scipy.stats import kendalltau
kendalltau(x,y)

SignificanceResult(statistic=-0.29814239699997197, pvalue=0.4205962375999266)

- (상관계수, p-value)
- 상관계수가 -0.298로 상관관계가 거의 없음

### 윌콕슨의 부호순위 검정

In [ ]:
# Q. 3절의 단일표본  t-검정 대신 일표본 부호순위 검정을 수행해보자.
from scipy.stats import wilcoxon

# 일표본 윌콕슨 부호순위 검정
# 몸무게 데이터 임의 생성
kg=np.array([75.5,83.9,75.7,56.2,73.4,67.7,79.0,50.7,58.4,74.1,65.1,77.8,48.1,46.3])

# 두 샘플의 크기가 다를 수 있으므로 길이를 맞추기
wilcoxon(kg-70) # 70으로 설정

WilcoxonResult(statistic=42.0, pvalue=0.5416259765625)

- (검정통계량, p-value)
- p-value > 0.05 귀무가설을 기각하지 못함, 중위수가 70과같다

In [18]:
# Q. 3절의 대응표본 t-검정 대신 이표본 부호순위 검정을 수행해보자
female=np.array([50.7,58.4,74.1,65.1,77.8,48.1,46.3])
male=np.array([75.5,83.9,75.7,56.2,73.4,67.7,79.0])

# 두 데이터의 차이
diff=female-male

# 윌콕슨 부호순위 검정 실시
wilcoxon(diff)

WilcoxonResult(statistic=5.0, pvalue=0.15625)

- (검정통계량, p-value)
- p-value > 0.05 귀무가설을 기각하지 못함, 두표본의 중앙값에 유의미한 ㅊ ㅏ이가 없다

### 윌콕슨의 순위합 검정

In [19]:
# Q. 3절의 독립표본 t검정 대신 순위합 검정을 수행해보자.
from scipy.stats import ranksums
ranksums(female,male)

RanksumsResult(statistic=-1.8527420384998257, pvalue=0.06391934147515746)

- p-value >0.05 귀무가설 기각 못함 두 표본의 중앙값에 유의미한 차이가 없다


### 만-위트니 U검정

In [20]:
# Q. '3절의 독립표본 t-검정' 대신 만-위트니 U 검정을 수행해보자.
from scipy.stats import mannwhitneyu
mannwhitneyu(female,male)

MannwhitneyuResult(statistic=10.0, pvalue=0.07284382284382285)

- p-value > 0.05 귀무가설을 기각하지 못함 . 두표본의 중앙값에 유의미한 차이가 없다

### 크루스칼-왈리스 H검정

In [ ]:
import pandas as pd
df=pd.read_csv('../예제/school_score.csv')

# 입력 형태에 맞게 데이터 추출
# 필요한 컬럼 각각 할당

School=df['School']
Final=df['Final']

# School에 대해 각각 기말고사 점수 분리
Final_A=Final[School=='A'].reset_index(drop=True)
Final_B=Final[School=='B'].reset_index(drop=True)
Final_C=Final[School=='C'].reset_index(drop=True)

# 크루스칼 - 왈리스 검정 수행
from scipy.stats import kruskal``
kruskal(Final_A,Final_B,Final_C)

KruskalResult(statistic=1978.1734792103248, pvalue=0.0)